### PR TITLE
Relocate a client window as if it is undecorated when reparenting it back

### PR DESCRIPTION
**Pull request by xinhaoyuan**
This helps resolve some annoying offsets in floating mode after closing and reopening some trayed applications, for example, Telegram. I'm not sure if this is breaking any protocols (I read some ICCCM but haven't found anything related yet).

### CODE REVIEW

**Comment by psychon** [comment]
From a quick look: Seems to be closely related to #2323 which I stopped working on (IIRC) since I couldn't get the tests to work on CI (they passed locally).

**Comment by xinhaoyuan** [comment]
Yes, that is the same issue I was trying to fix. I think my tests somehow works. Feel free to adopt anything to your PR if you think they can be useful.

**Review by Aire-One — APPROVED** [review]
I'm glad you submitted this patch! 

It looks even better than the path you linked on reddit as your `client_get_undecorated_geometry` function here also takes into account of the borders.

Thanks :)

(In the test, I can see you use `xwininfo`. I guess it should be added to the README next to `wmctrl` in the dependencies list)

File: tests/test-geometry.lua
Comment (psychon) [inline tests/test-geometry.lua:318]: Could you add a call to `awesome.sync()` afterwards? That should make sure that the X11 server *really*  already handled the `c:unmanage()` before we get to the next step.

File: tests/test-geometry.lua
Comment (psychon) [inline tests/test-geometry.lua:340]: What does this do? What do you mean with "make sure the window is removed"?

I *think*  that `wmctrl -c` doesn't do anything. According to https://github.com/dancor/wmctrl/blob/0181e8e4a7ed98255f6ef930875e56560f19aa95/main.c#L738-L740, this sends a `_NET_CLOSE_WINDOW` message to the root window. Such a message is handled by the WM. However, due to your `c:unmanage()` above, AwesomeWM does not know the window any more and I would expect this to do nothing.

**Review by psychon — COMMENTED** [review]
I'm not sure I understand the new tests.

- It adds a rule that moves the client to (0, 0) with size (100, 100), a border width of 10 and a titlebar of height 20 at the top
- It opens a test client with the gravity under test and tests that the rule was applied, then does `c:unmanage()`
- Afterwards you use `xwininfo` to try to figure out where the window ends up and check that the position is as expected for the gravity.
- Then you open another window, this time with a rule that enforces your "wanted" geometry via a rule and then test that the rule was correctly applied. This seems to be that it only tests `awful.rules` (or whatever it is called now), but does not test anything with gravity handling.
- Here, you also then explicitly test how the geometry changes due to the addition of the titlebar. Is this what you had in mind here? Was this broken before or why is this part of this PR? I'm a bit confused whether the "change itself" should fix this test or whether it is unrelated.

Also: Does this fix #2308?

File: tests/test-geometry.lua
Comment (xinhaoyuan) [reply tests/test-geometry.lua:316]: TIL, added.

File: tests/test-geometry.lua
Comment (xinhaoyuan) [reply tests/test-geometry.lua]: That's a really good point. I didn't pay much attention to how the close request is handled. I now uses another method to detect window existence - please let me know.

**Comment by xinhaoyuan** [comment]
Thanks for look into the PR! To answer your questions:

> (In the test, I can see you use xwininfo. I guess it should be added to the README next to wmctrl in the dependencies list)

I figured that I can use lgi instead of external programs, but I know sometimes it can cause issues. Currently the test works fine, but please let me know if there are actually issues with that.

> I'm not sure I understand the new tests.
> 
>     * It adds a rule that moves the client to (0, 0) with size (100, 100), a border width of 10 and a titlebar of height 20 at the top
> 
>     * It opens a test client with the gravity under test and tests that the rule was applied, then does `c:unmanage()`
> 
>     * Afterwards you use `xwininfo` to try to figure out where the window ends up and check that the position is as expected for the gravity.
> 
>     * Then you open another window, this time with a rule that enforces your "wanted" geometry via a rule and then test that the rule was correctly applied. This seems to be that it only tests `awful.rules` (or whatever it is called now), but does not test anything with gravity handling.

Yes, this is just to make sure the rule is applied. This step and the following step is to make sure that the initial expectation is "correct". So the entire steps checks that when a client is unmanaged then managed again, it is first relocated back to the expectation, then moved back to the initial geometry 100x100+0+0.

> 
>     * Here, you also then explicitly test how the geometry changes due to the addition of the titlebar. Is this what you had in mind here? Was this broken before or why is this part of this PR? I'm a bit confused whether the "change itself" should fix this test or whether it is unrelated.
> 
> 
> Also: Does this fix #2308?

I think it does - Telegram uses STATIC gravity, and closing and reopening it with the change does not change it position.

File: tests/test-geometry.lua
Comment (psychon) [reply tests/test-geometry.lua:316]: Since you write "TIL", let's actually explain my thinking here:
- `client_unmanage()` in `objects/client.c` "does some stuff" and sends requests to the X11 server.
- However, libxcb actually has an output buffer and does not send requests unless it actually has to.
- Even when the output buffer is flushed, that only means that the requests were sent, but not that they were handled by the X11 server.
- Thus, in theory it could be possible that  `wmctrl` manages to start and talk to the X11 server before the unmanage is actually done / handled by the X11 server.

The call to `awesome.sync()` sends a request to the X11 server and actually waits for the reply. X11 guarantees that the requests are handled in-order. Thus, by waiting for a reply to a new request, we can be sure that all the previous requests were handled.

**Comment by psychon** [comment]
Thanks for your comment. I think someone should run the test program from #2308 before this is merged. However, I do not have time right now. Sorry.

**Comment by xinhaoyuan** [comment]
I ran the code in #2308 and I think the issue is fixed.

But for some reason the CI checks in this PR seems to be not working..

**Comment by psychon** [comment]
> But for some reason the CI checks in this PR seems to be not working..

Sigh. The Travis web page says:

> Builds have been temporarily disabled for public repositories due to a negative credit balance. Please go to the Plan page to replenish your credit balance or alter your Consume paid credits for OSS setting.

If I understand this correctly, we cannot run builds until next month...?

**Comment by Elv13** [comment]
@xinhaoyuan Hello, I purchased some Travis credits to be able to run builds again. Mind making a random change to this to the CI sees it?

**Comment by actionless** [comment]
@Elv13 re-opening the PR also helps, will do it now

File: tests/test-geometry.lua
Comment (actionless) [inline tests/test-geometry.lua:264]: i guess this should be a comment, not a print statement

**Comment by actionless** [comment]
@xinhaoyuan test prev commits setup is failing in CI: https://travis-ci.com/github/awesomeWM/awesome/builds/222130810

please squash your commiits

**Comment by psychon** [comment]
I don't think this is `$TEST_PREV_COMMITS`. Luacheck also complains locally:
```
Checking tests/test-geometry.lua                  3 warnings

    tests/test-geometry.lua:327:25: shadowing upvalue 'w' on line 13
    tests/test-geometry.lua:366:27: unused variable 'geo'
    tests/test-geometry.lua:381:19: unused variable 'geo'

```

File: tests/test-geometry.lua
Comment (xinhaoyuan) [reply tests/test-geometry.lua:264]: Done

**Comment by xinhaoyuan** [comment]
I revised the patch with some simplification. I also realized that some default settings in awesomerc.lua are interfering with how new clients are placed/decorated. So I tried to override them in the tests and make them more robust.

**Comment by actionless** [comment]
CI is still failing

**Comment by xinhaoyuan** [comment]
> CI is still failing

Hmm.. I couldn't figure out what's wrong from the CI log. Any clues?

**Comment by psychon** [comment]
> Any clues?

More or less still the same `luacheck` complaint as in https://github.com/awesomeWM/awesome/pull/3253#issuecomment-813360765:
```
Checking tests/test-geometry.lua                  4 warnings

    tests/test-geometry.lua:304:27: unused variable 'geo'
    tests/test-geometry.lua:319:25: shadowing upvalue 'w' on line 13
    tests/test-geometry.lua:341:27: unused variable 'geo'
    tests/test-geometry.lua:363:19: unused variable 'geo'
```
(I have `luacheck` installed and can just run `make luacheck` to have luacheck run on the source code)

**Comment by xinhaoyuan** [comment]
> > Any clues?
> 
> More or less still the same `luacheck` complaint as in [#3253 (comment)](https://github.com/awesomeWM/awesome/pull/3253#issuecomment-813360765):
> 
> ```
> Checking tests/test-geometry.lua                  4 warnings
> 
>     tests/test-geometry.lua:304:27: unused variable 'geo'
>     tests/test-geometry.lua:319:25: shadowing upvalue 'w' on line 13
>     tests/test-geometry.lua:341:27: unused variable 'geo'
>     tests/test-geometry.lua:363:19: unused variable 'geo'
> ```
> 
> (I have `luacheck` installed and can just run `make luacheck` to have luacheck run on the source code)

Ah, nice find! I didn't have luacheck installed. Fixed.

**Comment by Aire-One** [comment]
Is there still any blocker for this to be merged?

For what it's worth, I just tried to use this branch and run the scenario from #2308:

- remove placement from rules
- build the C program with given gcc parameters
- start Xephir and run the C program

My observations were exactly what there is in the "expected" description:

> A 100×100 window should appear, the top left corner of the client area should be placed at (100, 100); the window should flash 5 times, but remain at the same position.

**Review by actionless — APPROVED** [review]

**Comment by Aire-One** [comment]
So, I'd like to approve this PR and give the opportunity to mergify to auto-merge it. But @Elv13 self-assigned it. Is there more to do in the scope of this PR?/Is there any remaining blocker?

**Comment by Elv13** [comment]
No, it was just so it was in my queue so I didn't forgot about it

**Review by Aire-One — APPROVED** [review]

**Pull request merged**
